### ISSUE TITLE
hidden state are not equal when or when not use FSDP wrap the mdel

### ISSUE THREAD

**Issue by hjc3613** [pytorch/pytorch#122859]
### 🐛 Describe the bug

I want to get the hidden states for 1k sentences using 70B model, I have 2 way to do it:
1、load the model in single process, with 3 nivdia A800, and iter the 1k sentence, select the hidden states I want and save it on dist using `safetensors.torch.save_file`
2、the 1th method is too slow and cost about 1h, so I decide to run the model using FSDP:
`
        
        ........
        mixed_precision_policy, wrapping_policy = get_policies(fsdp_config, rank)
        my_auto_wrapping_policy = fsdp_auto_wrap_policy(model, LlamaDecoderLayer)
        model = FSDP(
            model,
            auto_wrap_policy= my_auto_wrapping_policy if train_config.use_peft else wrapping_policy,
            cpu_offload=CPUOffload(offload_params=True) if fsdp_config.fsdp_cpu_offload else None,
            mixed_precision=mixed_precision_policy if not fsdp_config.pure_bf16 else None,
            sharding_strategy=fsdp_config.sharding_strategy,
            device_id=torch.cuda.current_device(),
            limit_all_gathers=True,
            sync_module_states=train_config.low_cpu_fsdp,
            param_init_fn=lambda module: module.to_empty(device=torch.device("cuda"), recurse=False)
            if train_config.low_cpu_fsdp and rank != 0 else None,
        )
        if fsdp_config.fsdp_activation_checkpointing:
            apply_fsdp_checkpointing(model)`

and launch the python code using `torchrun --nnodes 1 --nproc_per_node 8`

now I can run model with pipeline parallel and data parallel, the time cost down to 10min，Bug I also get some new problem，the hidden states of the same sentence are not equal：
![image](https://github.com/pytorch/pytorch/assets/37894838/1341775e-2891-4355-92fe-d29560083069)
Can anyone tell me te reason about it? and how dose it affect the distillation when student model learns the teacher model? 

node: the model.eval() has activate, the model is only in forward and no backward, no grad, no optimizer update

### Versions

the docker I using was: nvcr.io/nvidia/pytorch:23.10-py3

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang

**Comment by hjc3613** [comment]
I save the hidden states on disk only for model distillation, becuase I want to train a student model and learns the specific layer hidden states of teacher model, so I need save the teacher hidden states before distillation training

**Comment by hjc3613** [comment]
> Are you using pure bf16 for both single-GPU and FSDP?

yes

**Comment by hjc3613** [comment]
> It seems possible that using bf16 reduce-scatter with FSDP would cause some numerical differences compared to single GPU 🤔
> 
> On single-GPU, the gradient might be computed directly over the batch, whereas for FSDP, each rank has its local gradient over its local batch, and this must be summed across ranks. The reduction order might affect rounding and then change numerics.

thank you for your reply, but I just calculate the hidden states in forward, no backward, no grad, I have activated the model.eval() and with torch.no_grad(), is there something like round happen?

**Comment by awgu** [comment]
Can you clarify your setup more?

In particular, how are you maintaining the same global batch size across the single-GPU vs. multi-GPU FSDP setups? For example, if have 1k sentences globally, how are these being partitioned across workers and how are you aggregating them for the final hidden state?

**Comment by hjc3613** [comment]
> Can you clarify your setup more?
> 
> In particular, how are you maintaining the same global batch size across the single-GPU vs. multi-GPU FSDP setups? For example, if have 1k sentences globally, how are these being partitioned across workers and how are you aggregating them for the final hidden state?

in sinlge gpu, I just load model using transformers model.from_pretrained, it split model into 3 gpu (auto splited by transformers, I just using device_map='auto' and export CUDA_VISIBLE_DEVICES=0,1,2), to get the hidden states, I am using the bellow code:
`

batch = tokenizer(text, return_tensors='pt') 

output = model(**batch, output_hidden_states=True) 
all_hidden_states = output.hidden_states 
safetensors.torch.save(...) # save the layer hidden state wich I need 

`

the above method is too slow, so I want to utilize the pipeline parallel and data parallel provided by FSDP, so the second time, I wrap the model using the code:

`   
    mixed_precision_policy, wrapping_policy = get_policies(fsdp_config, rank)

    my_auto_wrapping_policy = fsdp_auto_wrap_policy(model, LlamaDecoderLayer)
    model = FSDP(
        model,
        auto_wrap_policy= my_auto_wrapping_policy if train_config.use_peft else wrapping_policy,
        cpu_offload=CPUOffload(offload_params=True) if fsdp_config.fsdp_cpu_offload else None,
        mixed_precision=mixed_precision_policy if not fsdp_config.pure_bf16 else None,
        sharding_strategy=fsdp_config.sharding_strategy,
        device_id=torch.cuda.current_device(),
        limit_all_gathers=True,
        sync_module_states=train_config.low_cpu_fsdp,
        param_init_fn=lambda module: module.to_empty(device=torch.device("cuda"), recurse=False)
        if train_config.low_cpu_fsdp and rank != 0 else None,
    )
    if fsdp_config.fsdp_activation_checkpointing:
        apply_fsdp_checkpointing(model)
`
now I can run the model on 8 gpu using the launch script  `torchrun --nnodes 1 --nproc_per_node 8`

the totoal code is much complex, but the core forward code I use was(the train_config.enable_fspt is True):
![image](https://github.com/pytorch/pytorch/assets/37894838/3ec1b04b-3daa-44a2-837d-745a9fe5ec18)
in the aboe code, the eval_dataloader was build by:
![image](https://github.com/pytorch/pytorch/assets/37894838/323e9676-42c0-4ef5-8f05-b0b34f4d0d27)

**Comment by hjc3613** [comment]
> > Can you clarify your setup more?
> > In particular, how are you maintaining the same global batch size across the single-GPU vs. multi-GPU FSDP setups? For example, if have 1k sentences globally, how are these being partitioned across workers and how are you aggregating them for the final hidden state?
> 
> in sinlge gpu, I just load model using transformers model.from_pretrained, it split model into 3 gpu (auto splited by transformers, I just using device_map='auto' and export CUDA_VISIBLE_DEVICES=0,1,2), to get the hidden states, I am using the bellow code: `
> 
> batch = tokenizer(text, return_tensors='pt')
> 
> output = model(**batch, output_hidden_states=True) all_hidden_states = output.hidden_states safetensors.torch.save(...) # save the layer hidden state wich I need
> 
> `
> 
> the above method is too slow, so I want to utilize the pipeline parallel and data parallel provided by FSDP, so the second time, I wrap the model using the code:
> 
> ` mixed_precision_policy, wrapping_policy = get_policies(fsdp_config, rank)
> 
> ```
> my_auto_wrapping_policy = fsdp_auto_wrap_policy(model, LlamaDecoderLayer)
> model = FSDP(
>     model,
>     auto_wrap_policy= my_auto_wrapping_policy if train_config.use_peft else wrapping_policy,
>     cpu_offload=CPUOffload(offload_params=True) if fsdp_config.fsdp_cpu_offload else None,
>     mixed_precision=mixed_precision_policy if not fsdp_config.pure_bf16 else None,
>     sharding_strategy=fsdp_config.sharding_strategy,
>     device_id=torch.cuda.current_device(),
>     limit_all_gathers=True,
>     sync_module_states=train_config.low_cpu_fsdp,
>     param_init_fn=lambda module: module.to_empty(device=torch.device("cuda"), recurse=False)
>     if train_config.low_cpu_fsdp and rank != 0 else None,
> )
> if fsdp_config.fsdp_activation_checkpointing:
>     apply_fsdp_checkpointing(model)
> ```
> 
> `now I can run the model on 8 gpu using the launch script `torchrun --nnodes 1 --nproc_per_node 8`
> 
> the totoal code is much complex, but the core forward code I use was(the train_config.enable_fspt is True): ![image](https://private-user-images.githubusercontent.com/37894838/317926275-3ec1b04b-3daa-44a2-837d-745a9fe5ec18.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE2NzYwNjEsIm5iZiI6MTcxMTY3NTc2MSwicGF0aCI6Ii8zNzg5NDgzOC8zMTc5MjYyNzUtM2VjMWIwNGItM2RhYS00NGEyLTgzN2QtNzQ1YTlmZTVlYzE4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzI5VDAxMjkyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJmNzE5NTA2YjBjNzBlODA4MmRjNDNmNTM4ZTRjN2ZjZmJkMzNmYjU0NTU0NGQ3ZjYxMjk5NmVlZjAyMmZmN2EmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Jkdjh-eFaRXVN_iskV8CbmAv2F1X3C9C1c48khMuOL4) in the aboe code, the eval_dataloader was build by: ![image](https://private-user-images.githubusercontent.com/37894838/317926856-323e9676-42c0-4ef5-8f05-b0b34f4d0d27.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE2NzYwNjEsIm5iZiI6MTcxMTY3NTc2MSwicGF0aCI6Ii8zNzg5NDgzOC8zMTc5MjY4NTYtMzIzZTk2NzYtNDJjMC00ZWY1LThmMDUtYjBiMzRmNGQwZDI3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzI5VDAxMjkyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc3YzJlZDA1ODczOTQxOGE3NTFiYzY1OWYwYTI5NzE3MDU3NjA1ODllOGNiY2MwN2UyODBhZWUyZDQ4M2NiMzkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.TumO-xvmZbCz9yF_1ELMJgkS2-PuWCSbkEeKEjBKxZo)

1、after get the output in fsdp method, I save the hidden states using the same code as in single process
2、also，the first time, I am using 3 gpu, but 1 process, the second time ,I am using 8gpu and 8process

**Comment by awgu** [comment]
I am not so familiar with `device_map="auto"`, but from [here](https://discuss.huggingface.co/t/device-map-auto/49610/3), it looks like it runs tensor parallelism across your 3 GPUs. In that case, are you using local batch size 8? (Under TP, all tensor parallel workers would work together on the same batch, so for the global batch size to match, each tensor parallel worker needs to process the global batch size of 8 as its local batch size.)

**Comment by hjc3613** [comment]
> I am not so familiar with `device_map="auto"`, but from [here](https://discuss.huggingface.co/t/device-map-auto/49610/3), it looks like it runs tensor parallelism across your 3 GPUs. In that case, are you using local batch size 8? (Under TP, all tensor parallel workers would work together on the same batch, so for the global batch size to match, each tensor parallel worker needs to process the global batch size of 8 as its local batch size.)

if set device_map='auto', the transformers can auto load the model on mutlple gpus, for example the 80 transformer layers for 70B model, 0-26 layers on gpu0, 27-52 layers on gpu1, 53-80 on gpu2, the embedding layer on gpu0 and lm head layer on gpu2, in short, the transformers can handle the model paralle perfectly. after split the model into 3gpu, it acts like on 1 gpu, we do not need care about the detail of tensor flow inner the model. there is no tensor parallel, only model parallel, the data flow in the model like :
![image](https://github.com/pytorch/pytorch/assets/37894838/e870c5a0-bae0-464d-8869-fd313f330e27)
the global batch size is 1

**Comment by awgu** [comment]
Thanks for the explanation! It looks like pipeline parallelism.

In that case, should you use global batch size 8 for the `device_map="auto"` case to make it comparable with 8-way FSDP with local batch size of 1?

**Comment by hjc3613** [comment]
> Thanks for the explanation! It looks like pipeline parallelism.
> 
> In that case, should you use global batch size 8 for the `device_map="auto"` case to make it comparable with 8-way FSDP with local batch size of 1?

I think the batch size of 1 in single process case is comparable with 8-way FSDP, the reasons are:
1、in FSDP, each gpu handle only 1 text(micro batch size is 1), is comparable with batch size of 1 in single process case,  there is no padding in reality(in my dataloader, only if micro batch size > 1, it padding to the same length)
2、if use batch size of 8 in single process, I must pad the input ids to equal since different length of some 8 text, this can be not comparable to 8-way FSDP

**Comment by awgu** [comment]
To make sure I am understanding, in the single-GPU case, are you accumulating gradients over 8 batch elements before running optimizer step? In 8-way FSDP, when each GPU computes with local batch size 1, the GPUs will reduce-scatter gradients such that they have the appropriate shard of the reduced gradients for 8 batch elements.

**Comment by awgu** [comment]
Sorry for misunderstanding -- this makes sense.

I think at this point, if you want to rule out floating-point associativity differences, you can run both with fp32 instead of bf16, if your GPU memory allows for it. If the difference is much smaller, then it might be an floating-point associativity issue.

**Comment by hjc3613** [comment]
> Sorry for misunderstanding -- this makes sense.
> 
> I think at this point, if you want to rule out floating-point associativity differences, you can run both with fp32 instead of bf16, if your GPU memory allows for it. If the difference is much smaller, then it might be an floating-point associativity issue.

thanks for your help, I will try it

**Comment by awgu** [comment]
Closing for now since we have some general resolution (try fp32) -- feel free to reopen if there are follow-ups
